src/magicbox/QueryProcessor.ts

```diff
@@ -0,0 +1,99 @@
+export interface IQueryProcessorOptions {
+  timeout: number;
+}
+
+export enum ProcessingStatus {
+  Finished,
+  TimedOut,
+  Overriden
+}
+
+export interface IQueryProcessResult<T> {
+  status: ProcessingStatus;
+  results?: T[];
+}
+
+/**
+ * IE11 equivalent of Promise.race
+ */
+function racePromises<T>(...promises: Thenable<T>[]): Promise<T> {
+  let done = false;
+  return new Promise<T>((resolve, reject) => {
+    const finish = (func: Function) => {
+      if (done) {
+        return;
+      }
+      done = true;
+      func();
+    };
+    promises.forEach(promise => promise.then(result => finish(() => resolve(result))).catch(err => finish(() => reject(err))));
+  });
+}
+
+export class QueryProcessor<T> {
+  private override: () => void;
+  private options: IQueryProcessorOptions;
+  private processedResults: T[];
+
+  constructor(options: Partial<IQueryProcessorOptions> = {}) {
+    this.options = { timeout: 500, ...options };
+  }
+
+  /**
+   * Overrides the previous queries and accumulates the result of promise arrays with a timeout.
+   */
+  public async processQueries(queries: (T[] | Promise<T[]>)[]): Promise<IQueryProcessResult<T>> {
+    this.overrideIfProcessing();
+    this.processedResults = [];
+    const asyncQueries = queries.map(query => (query instanceof Promise ? query : Promise.resolve(query)));
+
+    return racePromises(
+      this.resolveQueriesAndAccumulateResults(asyncQueries).then(() => this.buildProcessResults(ProcessingStatus.Finished)),
+      this.waitForOverride().then(() => this.buildProcessResults(ProcessingStatus.Overriden)),
+      this.waitForTimeout().then(() => this.buildProcessResults(ProcessingStatus.TimedOut))
+    );
+  }
+
+  public async overrideIfProcessing() {
+    if (this.override) {
+      this.override();
+    }
+  }
+
+  private buildProcessResults(status: ProcessingStatus): IQueryProcessResult<T> {
+    return {
+      status,
+      ...(this.statusHasResults(status) && { results: this.processedResults })
```

Comment:
This was just because `results` was optional and I didn't want to pass it any value when `statusHasResults` was `false`.

Now that `results` is non-optional, I'll remove `statusHasResults` and change this to just pass `this.processedResults`.